crates/flipperzero/src/furi/log/metadata.rs

```diff
@@ -0,0 +1,535 @@
+//! Metadata describing log data.
+//
+// The structs and enums in this file are extracted from the `tracing-core` crate with
+// adaptions to Furi. The original code is copyright (c) 2019 Tokio Contributors
```

Comment:
This seems a lot of code for what should mostly be just some macros around `sys::furi_log_print_format`.

The [`log`](https://crates.io/crates/log) crate has `no-std` support and [only needs about 15 lines to implement](https://docs.rs/log/0.4.17/log/#implementing-a-logger). As a bonus, we'd immediately support any crate that used the `log` crate as a logging facade.

The [`tracing`](https://crates.io/crates/tracing) crate is quite powerful, but the main reason for using it is to be able to understand asynchronous code where tasks can jump between threads. While it looks like it could be possible to implement a [`tracing::Subscriber`](https://docs.rs/tracing/0.1.37/tracing/trait.Subscriber.html), there's a lot of code that needs to be generated to support spans, which I worry would blow out binary sizes quite quickly.

Reply:
I actually started this effort by trying to write a `log` backend, and I very quickly ran into the problem that `log` assumes that you will be writing and formatting with `&str`, while `sys::furi_log_print_format` requires `&CStr`. This would force inline allocations at every individual log point, in order to copy the format string and append a `nul` byte.

`tracing` has the same issue here. It's a fantastic library and I use it wherever I can (it's useful for way more than just understanding async code), but it also is built around `&str`, and thus doesn't match up well with `sys::furi_log_print_format`.

My next thought was that we can use the `c_string!` macro to ensure that the format string is statically null-terminated, but this requires writing our own macros. Once we've done that, the next step is to represent the `FuriLevel` enum on the Rust side, which is where I turned to `tracing`'s implementation as it is heavily optimised for low code overhead, which we want here.

Reply:
All that being said: if we're going to use `ufmt` and `FuriString` for formatting log output before calling `sys::furi_log_print_format`, then we _could_ provide a `log::Log` and/or `tracing::Subscriber` implementation that users can choose to install.

Reply:
I think `FuriString` is going to be the right approach. We _may_ be able to do something clever like using a static `FuriString` buffer for formatting, so we're not constantly doing allocations (the logger is already [serialized on a mutex](https://sourcegraph.com/github.com/flipperdevices/flipperzero-firmware@61744b8d01e7ad1d7544a6a0176fac5369e67dc4/-/blob/furi/core/log.c?L27)).

I just realized another issue with the `log` crate. It expects [records to be formatted as `core::fmt::Arguments`](https://docs.rs/log/0.4.17/log/struct.Record.html#method.args) which pretty much forces the use of `core::fmt` which we're trying to avoid. Tracing is a bit better since it can keep track of individual fields, but generic messages are still formatted using `core::fmt`.

Reply:
We now use `FuriString` (allocating each time), and `ufmt` for formatting. So our arguments are whatever `ufmt::uwrite` takes instead of `core::fmt::Arguments`.

We could maybe use the `rt` macros to set up a static, and then free it on exit. That seems like an approach that would be generally useful for various sub-systems; opened #64.